shiny/ui/_theme_brand.py

```diff
@@ -278,35 +195,30 @@ def __init__(
         )
 
         self.brand = brand
+        self.add_sass_layer_file(
+            path_pkg_www("..", "py-shiny", "brand", "_brand-yml.scss")
+        )
 
         # Prep Sass and CSS Variables -------------------------------------------------
-        sass_vars_theme_colors, sass_vars_brand_colors, css_vars_brand = (
-            ThemeBrand._prepare_color_vars(brand)
+        (
+            brand_color_palette_defaults,
+            brand_color_defaults,
+            brand_color_palette_rules,
+        ) = ThemeBrand._prepare_color_vars(brand)
+
+        brand_typography_defaults = ThemeBrand._prepare_typography_vars(brand)
+
+        brand_bootstrap_defaults = (
+            "\n".join(Theme._combine_args_kwargs(kwargs=brand_bootstrap.defaults))
+            if brand_bootstrap.defaults
+            else ""
         )
-        sass_vars_typography = ThemeBrand._prepare_typography_vars(brand)
-
-        # Theme -----------------------------------------------------------------------
-        # Defaults are added in reverse order, so each chunk appears above the next
-        # layer of defaults. The intended order in the final output is:
-        # 1. "Brand" Color palette
-        # 2. "Brand" Bootstrap Sass vars
-        # 3. "Brand" theme colors
-        # 4. "Brand" typography
-        # 5. Gray scale variables from "Brand" fg/bg or black/white
-        # 6. Fallback vars needed by additional "Brand" rules
-
-        self.add_defaults("", "// *---- brand: end of defaults ----* //", "")
-        self._add_sass_ensure_variables()
-        self._add_sass_brand_grays()
-        self._add_defaults_hdr("typography", **sass_vars_typography)
-        self._add_defaults_hdr("theme colors", **sass_vars_theme_colors)
-        if brand_bootstrap.defaults:
-            self._add_defaults_hdr("bootstrap defaults", **brand_bootstrap.defaults)
-        self._add_defaults_hdr("brand colors", **sass_vars_brand_colors)
-
-        # "Brand" rules (now in forwards order)
-        self._add_rules_brand_colors(css_vars_brand)
-        self._add_sass_brand_rules()
+
+        self._insert_sass("brand.color.palette:defaults", brand_color_palette_defaults)
+        self._insert_sass("brand.defaults:defaults", brand_bootstrap_defaults)
+        self._insert_sass("brand.color:defaults", brand_color_defaults)
+        self._insert_sass("brand.typography:defaults", brand_typography_defaults)
+        self._insert_sass("brand.color.palette:rules", brand_color_palette_rules)
```

Comment:
Any chance this could be done with `.add_defaults()` and `.add_rules()` instead of `._insert_sass()`?

Reply:
I also like that you can look at the Sass file and have a sense of the pieces that are going to be added. 

Also string substitution isn't elegant but it's certainly easy to do in both Python and R and both implementations will basically be the same thing. It's likely that using the Sass layer mechanisms will work out to look pretty similar in both languages, but it certainly increases the context window of things you need to understand.

In other words, I'm going for a solution that, as much as possible, fits into one Sass file.

Reply:
Anyway, that's what I was thinking but I rewrote it to use standard Sass methods and I don't hate it. 4ddf0f51451f98529b8da4785972e587e24537ee